core/conn/unixodbc/odbc/odbcclient/unixcli/cli/ctosqlconv.cpp

```diff
@@ -157,852 +157,982 @@ unsigned long ODBC::Ascii_To_Interval_Helper(char *source,
+        CDescRec*     targetDescPtr,
+        ICUConverter* iconv,
+        SQLPOINTER    targetDataPtr, 
+        UCHAR         *errorMsg)
+{
+    SQLSMALLINT    ODBCDataType    = targetDescPtr->m_ODBCDataType;
+    SQLSMALLINT    SQLDataType     = targetDescPtr->m_SQLDataType;
+    SQLSMALLINT    targetScale     = targetDescPtr->m_ODBCScale;
+    SQLSMALLINT    targetUnsigned  = targetDescPtr->m_SQLUnsigned;
+    SQLINTEGER     targetPrecision = targetDescPtr->m_ODBCPrecision;
+    SQLINTEGER     targetCharSet   = targetDescPtr->m_SQLCharset;
+    SQLINTEGER     translateLength = 0;
+    SQLINTEGER     targetLength    = targetDescPtr->m_SQLOctetLength;
+    SQL_INTERVAL_STRUCT *intervalTmp  = NULL;
+    char           srcDataLocale[CHARTMPLEN] = {0};
+    int            tempLen         = 0;
+    int            numberLen       = 0;
+    CHAR           cTmpBuf[CHARTMPLEN]    = {0};
+    CHAR           cTmpBuf2[CHARTMPLEN]    = {0};
+    BOOL           dataTruncatedWarning = FALSE;
+    unsigned long  retCode         = SQL_SUCCESS;
+    double         dTmp            = 0;
+    __int64        integralPart    = 0;
+    __int64        decimalPart     = 0;
+    SQLINTEGER     DataLen         = DRVR_PENDING;
+    SQLPOINTER     DataPtr         = NULL;
+    BOOL           useDouble       = TRUE;
+    long           leadZeros       = 0;
+    unsigned __int64  integralMax       = 0;
+    unsigned __int64  decimalMax        = 0;
+    BOOL           negative        = FALSE;
+    long           decimalDigits   = 0;
+
+
+    if(SQLDataType == SQLTYPECODE_NUMERIC || SQLDataType == SQLTYPECODE_NUMERIC_UNSIGNED)//for bignum support
+    { //Bignum
+        switch (CDataType)
+        {
+            case SQL_C_DEFAULT:
+                if (ODBCAppVersion >= SQL_OV_ODBC3)
+                {
+    
+                }                       // Want it fall thru and treat it like SQL_C_CHAR
             case SQL_C_WCHAR:
                 if (iconv->isAppUTF16())
                 {
                     if (srcLength != SQL_NTS)
                         srcLength = srcLength/2;
+                    // translate from UTF16
                     if (iconv->WCharToUTF8((UChar*)srcDataPtr, srcLength, srcDataLocale, sizeof(srcDataLocale), (int*)&translateLength, (char*)errorMsg) != SQL_SUCCESS)
                         return IDS_193_DRVTODS_ERROR;
                     srcDataPtr = srcDataLocale;
                     srcLength = translateLength;
                 }
+            case SQL_C_BINARY:
             case SQL_C_CHAR:
-                retCode = ConvertCharToNumeric(srcDataPtr, srcLength, dTmp);
-                if (retCode != SQL_SUCCESS)
-                    return retCode;
+                if (srcLength == SQL_NTS)
+                    tempLen = strlen((const char *)srcDataPtr);
+                else
+                    tempLen = srcLength;
+                if (tempLen > CHARTMPLEN - 1)
+                    strncpy(cTmpBuf, (char *)srcDataPtr, CHARTMPLEN - 1);
```

Comment:
Shouldn't we add a null terminator? I don't see tempLen being passed to Ascii_To_Bignum_Helper below so I assume it is depending on the presence of a null terminator.

Reply:
Do you mean that tempLen should be set 0 after use?

Reply:
No, I mean adding a line "cTmpBuf[CHARTMPLEN-1] = '\0';" after the strncpy. The "if" at 723 guarantees that no trailing null character is copied, so it seems like we need to add one if we take the code path of statement 724.

Reply:
Since I have set the all the bytes to '\0' by "CHAR cTmpBuf[CHARTMPLEN] = {0};", will the function strncpy change the value of cTmpBuf[CHARTMPLEN-1] ?

Reply:
Ah, you are correct. Your initialization sets all the bytes in the array to zero. So, no additional code to null-terminate is needed here.